### DISCUSSION TITLE
Block specific ports

### DISCUSSION THREAD

**Discussion by vladosam** [friendly-bits/geoip-shell#2] [General]
Hi, is there option to geoblock some ports and leave other ports unblocked?

**Comment by friendly-bits** [comment]
> Hi, is there option to geoblock some ports and leave other ports unblocked?

It is possible with nftables but currently my code doesn't implement this. You can add a feature request and if there will be more people asking for this feature, I'll implement it.

In the meanwhile, you could modify the geoip-shell-apply.sh script to block some ports. The code is reasonably well commented, so you should be able to find the relevant lines easily. If you need more specific directions, please describe your use case and what you want to block and allow (open all ports except what's closed? close all ports except what's open? which protocols?)

**Reply by vladosam** [reply]
Thanks for reply. I want to geoblock certain ports and allow all others. For example geoblock: TCP: 21115-21119 and UDP: 21116. I use traefik geoblock plugin for 80 and 443 ports but traefik can only geoblock HTTP-S ports.

**Reply by friendly-bits** [reply]
Which mode do you want to use? Whitelist or blacklist? Whitelist blocks all countries except the ones you allow, blacklist allows all countries except the ones you block.

**Reply by vladosam** [reply]
Whitelist.

**Reply by friendly-bits** [reply]
I'll look into how to modify the rules for your use case, hopefully will get back to you soon'ish. If you want to check this out by yourself, read nftables documentation, for example
https://wiki.nftables.org/wiki-nftables/index.php/Quick_reference-nftables_in_10_minutes
https://wiki.nftables.org/wiki-nftables/index.php/Simple_ruleset_for_a_server
https://www.netfilter.org/projects/nftables/manpage.html

**Reply by vladosam** [reply]
Thanks.

**Reply by friendly-bits** [reply]
Try this:
0. if geoip-shell is currently installed on your machine, uninstall it first
1. download the latest release
2. (before installation) edit the file geoip-shell-apply.sh
3. On line 247, (just before `# established/related`), copy-paste the following 2 lines:

```
	printf '%s\n' "insert rule inet $geotable $geochain tcp dport != 21115-21119 counter accept comment ${geotag_aux}_custom"
	printf '%s\n' "insert rule inet $geotable $geochain udp dport != 21116 counter accept comment ${geotag_aux}_custom"
```

`counter` is optional, the comment is necessary. You can modify the ports if you need to, or add more rules following the same pattern.

4. Save the file
5. Run the install script.
6. After installation, run `sudo geoip-shell status -v` and make sure that the custom rules are included
7. Verify that everything works as you want.

Please let me know if this worked or not.

Note that this is a custom-tailored solution for your specific use case as you described it. If you update the suite in the future, you will need to edit the updated script again and to make sure that the custom rules still fit into the updated code (which I can not guarantee in advance). Rules order matters with nftables.

**Reply by vladosam** [reply]
I'll try this tomorrow on my Debian 12 server and get back to you. Will this tool insert rules in iptables? What happens with rules if I uninstall it?

**Reply by friendly-bits** [reply]
This version only works with nftables. See the pre-requisites section. An earlier version worked with iptables but I moved on from developing it and it's not really ready for public consumption. Probably I will eventually make it available publicly but not right now.
That said, I see that Debian 12 supports nftables and probably you already have it installed. Check with `sudo nft --version`.

As to rules, they are all created in a new table, which is called 'geoip-shell'. The uninstall script removes that table completely (and any other geoip-shell files in the system). All this is written in the documentation btw.

**Reply by friendly-bits** [reply]
"nftables is the default framework in use in Debian (since Debian 10 Buster)"
https://wiki.debian.org/nftables

**Reply by vladosam** [reply]
I will test it tomorrow but I also know that iptables are installed on that server. I will see tomorrow how it goes.

**Reply by friendly-bits** [reply]
you probably have iptables-nft, which is a compatibility layer which accepts iptables syntax but actually uses nftables.
run `sudo iptables -V`, you will probably see something like `iptables v1.7 (nf_tables)`
https://manpages.debian.org/bookworm/iptables/iptables-nft.8.en.html

**Reply by vladosam** [reply]
No. It uses legacy. iptables v1.8.9 (legacy)

**Reply by friendly-bits** [reply]
Then you are out of luck, unless you want to replace iptables-legacy with nftables/iptables-nft. Trying to use 2 different frameworks simultaneously is a bad idea.

**Reply by friendly-bits** [reply]
I uploaded the iptables branch of the project. You can geoblock specific ports with it this way. Follow the directions from earlier, except:
In (1) download the release of the -iptables branch.
In (3), copy-paste the following 2 lines at line 307 (just before `#established/related`)
```
		printf '%s\n' "-I $geochain -p tcp --dport 21115:21119 -j ACCEPT -m comment --comment ${geotag}_aux_custom"
		printf '%s\n' "-I $geochain -p udp --dport 21116 -j ACCEPT -m comment --comment ${geotag}_aux_custom"
```

**Reply by vladosam** [reply]
Sorry I didn't reply but I have been ill for the last few days. I can't try that right now on my server. I don't feel comfortable to convert from iptables-legacy to iptables-nft on this server. I have some services like wireguard who sets iptables-legacy rules. But I will try this when I set new server. You should post geoip-shell on Reddit r/selfhosted. I think people there will like this very much.

**Reply by friendly-bits** [reply]
Thanks! Anyway, the iptables branch doesn't require iptables-nft and should work with iptables-legacy. I'm using it on my Debian Stretch server which doesn't have nftables. Check out in the releases, there is a version for iptables.

**Reply by friendly-bits** [reply]
Initially I developed this project for my Debian server and iptables. Later I reworked it for nftables, added new features, fixed bugs etc. So the iptables branch was left behind in development, which is why I told you that it's not ready for public consumption. In the past few days I patched it up to be on par with the nftables branch.

**Reply by friendly-bits** [reply]
The documentation for the iptables branch is here:
https://github.com/blunderful-scripts/geoip-shell/tree/geoip-shell-iptables

**Reply by vladosam** [reply]
Ok. I will test it than. Thanks.

**Reply by vladosam** [reply]
I set everything up as per your instructions.  As I said before, I have traefik and wireguard on this server. What will geoip-shell do with 22,80,443 and 51820 ports?

**Reply by friendly-bits** [reply]
Oh wait, I just realized that the additional rules I suggested you are incorrect. These are the (I believe) correct rules:

```
		printf '%s\n' "-I $geochain -p tcp ! --dport 21115:21119 -j ACCEPT -m comment --comment ${geotag}_aux_custom"
		printf '%s\n' "-I $geochain -p udp ! --dport 21116 -j ACCEPT -m comment --comment ${geotag}_aux_custom"
```

The difference is the `!` before `--dport`.
Basically these rules say "check if destination port is ..., if it's not then `ACCEPT`, i.e. don't geoblock, otherwise pass the packet to the geoblocking rules". Meaning that traffic arriving on other ports will not be subjected to geoblocking rules.

**Reply by friendly-bits** [reply]
You should test that this actually does exactly that.

**Reply by vladosam** [reply]
Excellent. I will try it today.

**Reply by friendly-bits** [reply]
I should note that what the ACCEPT verdict does is it makes the ACCEPTed packets bypass all other rules in the same table it was issued. The packet then continues to traverse further tables, and may be dropped in any of them. The nftables branch of geoip-shell creates a separate table and all geoip blocking is done within it. The iptables branch doesn't do that because of limitations of iptables. Instead, the iptables branch uses the existing **mangle** table and creates custom chains within it, attaching to the existing PREROUTING chain. What all this means essentially is that if you have further rules in later tables, for example in the **filter** table then they may subject the packets to additional rules.

**Comment by vladosam** [comment]
I'm getting ipv4 address '98.140.240.90' failed regex validation. I change that ip for security purpose. I want to use only my public address. 
```
Autodetected ipv4 LAN subnets: '172.17.0.0/16 172.20.0.0/16 172.21.0.0/16 10.0.0.0/24 10.20.30.2/32 98.140.240.90/32'.
(c)onfirm, c(h)ange, (s)kip or (a)bort installation? 
c|h|s|a: 

Type in ipv4 LAN subnets (whitespace separated), or Enter to abort installation.
98.140.240.90/32
ipv4 address '98.140.240.90' failed regex validation.
```

**Reply by friendly-bits** [reply]
Yes, sorry about that - I forgot to source the ip-regex script in the iptables version of the -install script. Just fixed it about an hour ago - please download the new version.

**Reply by friendly-bits** [reply]
Also, I'm a bit confused by you saying that this is your public address. The question asked by the installer is about the **LAN** (a.k.a local area network) subnets. Subnets you type in should be trusted subnets which will be *excluded* from geoip blocking. The idea is simply to prevent situation where you get locked out of your machine on your local area network.

**Reply by friendly-bits** [reply]
And also, if it's your public address then I don't understand how the autodetect code could find it. Unless this machine acts as a router or has otherwise direct connection to the provider? In both of these situations, your machine would have a network interface dedicated to connection with WAN (wide-area network) and if so then you should install the suite for router as in this mode, the suite will apply geoblocking rules to the specific WAN interface or interfaces and avoid processing any connections from the other (presumably LAN) interfaces.

**Reply by vladosam** [reply]
Aha. Ok. This is vps server. It's good thing than that it failed to install. :D I will whitelist only internal networks and leave public ip out of it.

**Reply by vladosam** [reply]
It seems that it works. Yeeeee. When I vpn to blacklisted country I can't connect. Also I tried `geoip-shell off` and than I can connect from blacklisted country. Very nice. Thanks for your support and don't forget to post on selfhosted subreddit.

**Reply by friendly-bits** [reply]
Great, happy to help. I don't really use reddit. If you feel that this project deserves a mention by you then I'll appreciate this of course.

**Reply by friendly-bits** [reply]
I probably need to look into better supporting VPS, somehow this never crossed my mind.

**Reply by friendly-bits** [reply]
BTW, since the autodetect code incorrectly thinks that your WAN subnet belongs to your LAN, make sure that you disabled autodetection when installing. Otherwise, the autodetect code will whitelist your WAN subnet during the next autoupdate. Then again, I'd recommend you to install geoip-shell in router mode. This way, instead of whitelisting LAN subnets, it only applies geoip blocking to the WAN interface. All that assuming that there is a dedicated WAN interface in your VPS server.

**Reply by friendly-bits** [reply]
If using geoip-shell in _host_ mode then specifically, when the installer asks:
`(A)uto-detect local subnets when autoupdating and at launch or keep this config (c)onstant?`
you should answer `c` to keep the config constant.

**Reply by vladosam** [reply]
I did choose that constant option. It looked logical to set it that way.

**Reply by friendly-bits** [reply]
In the latest release, I improved the wording of the explanation concerning WAN interfaces. Now this is no longer called 'router mode' but rather simply asks whether the machine has any WAN interfaces. Hopefully this will make it less confusing. The status report in the new version for iptables also provides statistics about how much each rule has been "hit" by the traffic.

Also If you don't mind, I'd like to find out why the LAN autodetection code incorrectly detects your WAN subnet. I would need a few technical details about the network on that machine.

**Reply by vladosam** [reply]
I'll help you if I can. What do you need to know?

**Reply by vladosam** [reply]
Now I can see if my block rule actually block something?

**Comment by friendly-bits** [comment]
> I'll help you if I can. What do you need to know?

I'll need output of these 3 commands (the latter 2 present more or less the same information but with slightly different detail):

`sudo ip -f inet route show table local scope link`

`sudo ip -f inet route show table local scope link | grep -v "[[:space:]]lo[[:space:]]" | grep -oE "dev[[:space:]]+[^[:space:]]+" | sed 's/^dev[[:space:]]*//g' | sort -u | while read -r iface; do sudo ip -o -f inet addr show "$iface"; done`

`sudo ip -f inet route show table local scope link | grep -v "[[:space:]]lo[[:space:]]" | grep -oE "dev[[:space:]]+[^[:space:]]+" | sed 's/^dev[[:space:]]*//g' | sort -u | while read -r iface; do sudo ip -f inet addr show "$iface"; done`

**Reply by friendly-bits** [reply]
Thank you. I copied the info, you can remove this now, for privacy.

**Reply by friendly-bits** [reply]
It appears that `eth0` is your dedicated WAN interface. Hopefully, if you run the installation again (with the new version) and answer `y` to the question about WAN interfaces, it should autodetect it correctly.

**Reply by vladosam** [reply]
Yes eth0 is WAN interface. I'll try updated version tomorrow.

**Reply by vladosam** [reply]
Also I write a post on reddit about geoip-shell. Hopefully it will attract some new users for you.

**Reply by friendly-bits** [reply]
So the issue with the LAN subnets autodetection code is that it has to use a heuristic because there is no guaranteed method of discerning between LAN and WAN subnets. The best heuristic that I found was looking in the routing table and checking which connections have scope 'link' (rather than scope 'host'). This is what the first command is doing. I believe that this will always or almost always work on a machine behind a router. This also works on my OpenWrt router. But in the case of your VPS, for some reason the WAN subnet is defined with 'scope link', which is why my code thinks that it's a LAN subnet. Not sure I have anything to do about this.

**Reply by friendly-bits** [reply]
(but then, the LAN subnets autodetection code is not supposed to be used on a machine which has direct WAN connection)

**Reply by vladosam** [reply]
What if script asks: is this vps server and than ask for correct WAN interface? Than you can ignore LAN autodetection completely.

**Reply by friendly-bits** [reply]
This is exactly what the install script is doing already. It will not ask you about LAN subnets if you answer 'y' in the question about WAN interfaces.

**Reply by friendly-bits** [reply]
(and LAN subnets will be ignored in this case)

**Reply by vladosam** [reply]
Mhm then maybe just some additional info on that question that if you use VPS WAN and LAN interface can be the same. I'm just spitballing. And I don't know if this makes any sense at all. :)

**Reply by friendly-bits** [reply]
I don't know that either. Besides, the target audience of this project is expected to have at least basic understanding of firewall-related concepts. One of my main goals is to make installation easy but making it completely fool-proof is not in the scope of this project. If the user doesn't understand that when asked about WAN interface, they need to provide correct answer, and doesn't read the installation guide which specifically warns about the importance of providing the correct answer then well, too bad for them.

**Reply by vladosam** [reply]
It's you call. Anyway i installed new version. Selected Y for Does this machine have dedicated WAN interface(s)? Script recognized eth0 as WAN interface and that was it. No more questions about lan subnets.

**Reply by friendly-bits** [reply]
Great. And thanks for the Reddit post! It is kinda hard for a small and niche project to get exposure and this definitely helps.

**Reply by vladosam** [reply]
You're welcome. Is there any chance to add rules with command such as: geoip-shell tcp 21115-211116 -c <"country_codes"> or something like that? That would be great feature.

**Reply by friendly-bits** [reply]
I'll look into how to implement this properly.

**Reply by friendly-bits** [reply]
I implemented your request. The syntax is a little odd but I'm sure you'll manage.

**Reply by friendly-bits** [reply]
New version 0.1.9 implements a more user-friendly syntax for protocols/ports setting, allows for separate control over each protocol, and the status report now includes information about protocols and ports geoblocking config. (I'm not setting the iptables version as the latest release but you'll find it inside Releases)

**Reply by vladosam** [reply]
Hi. I want to install new 0.1.9 and now I'm looking how to replicate current setup. So before we had to add this code in geoip-shell-apply.sh 
```
# rustdesk geoip
printf '%s\n' "-I $geochain -p tcp ! --dport 21115:21119 -j ACCEPT -m comment --comment ${geotag}_aux_custom"
printf '%s\n' "-I $geochain -p udp ! --dport 21116 -j ACCEPT -m comment --comment ${geotag}_aux_custom"
```
And I installed it with `sh geoip-shell-install.sh -m whitelist -c "DE"`. And that worked as intended. Only traffic from DE on tcp 21115-21119 and udp 21116 is checked for geoip and all other ports are unchecked. 
For new version.
First I need to remove old version. `geoip-shell-uninstall.sh -r`
Now with new version I believe command is 
```
sh geoip-shell-install -c "DE" -m whitelist -p tcp:block:21115-21119
sh geoip-shell-install -c "DE" -m whitelist -p udp:block:21116
```
And I dont need to add anything in  geoip-shell-apply.sh?

**Reply by friendly-bits** [reply]
You don't have to specifically call the uninstaller as the install script calls it anyway. If you want to be on the "safe-safe-safe" side, call the uninstall script without the `-r` option.

For the installation, you can do it in one command:
`sh geoip-shell-install -c "DE" -m whitelist -p tcp:block:21115-21119 -p udp:block:21116`

**Reply by friendly-bits** [reply]
If you need to change the ports later, you can do it this way:
`geoip-shell apply -p tcp:block:[ports] -p udp:block:[ports]`

**Reply by friendly-bits** [reply]
And yes, you no longer need to modify the apply script.

**Reply by vladosam** [reply]
It works like a charm. :tada: And it is super easy to set ports now.

**Reply by friendly-bits** [reply]
Thanks for the update, makes me very happy to hear that.

**Reply by vladosam** [reply]
Now that you have reddit account you should post updates about project every 2-3 months. With changelogs from last post to latest version. People on selfhost subreddit don't like when you flood them with every update but occasional posts are fine.

**Reply by friendly-bits** [reply]
My current focus is on making the project compatible with OpenWrt. At this point I think I just need to make a startup script for it (because on OpenWrt cron-based persistence doesn't work). Once that is complete, I'll announce it on their forum. I hope this will increase project's visibility. I'll probably write a post on Reddit as well at some point.

**Reply by friendly-bits** [reply]
I think I may also need to change my github name to something that sounds more trustworthy than blunderful-scripts, hehe. In the beginning my scripts had actually a ton of "blunders" (a.k.a. mindless mistakes) and were generally pretty bad but I improved a lot since then and the scripts improved a lot, so at this point this name probably makes a bad impression which is unwarranted.

**Reply by vladosam** [reply]
I didn't know what blunderful means but now that you said it changing name sounds like a good idea. :)

**Reply by friendly-bits** [reply]
aaaand here is the new username
(and OpenWrt support is complete for the nftables branch but now I want to merge the 2 branches before announcing it)

**Comment by friendly-bits** [comment]
> Now I can see if my block rule actually block something?

You'll be able to see statistics of traffic hitting each rule, including the whitelist blocking rule and the ACCEPT rules. So in short, yes. To see that, use the `geoip-shell status -v` command to have a verbose report, and there to the left of each rule you'll see the statistics.

If you update, don't forget to re-add your custom rules to the -apply script (now it's line 306).

**Comment by friendly-bits** [comment]
@vladosam I wonder if you are still using the project and what your experience with it is. Also you probably have already noticed, but just to make sure: geoip-shell doesn't have separate branches for iptables and nftables anymore. All recent releases support both firewall backends (and I implemented the option to change the backend if you have both nftables and iptables-compat installed). There were a few bugs affecting iptables installs because the code went through substantial changes, now I believe all of them have been ironed out.

P.s. I tried to write a post on the selfhosted subreddit but it didn't get approved by the admins, probably because my karma is not good enough xD.

**Reply by vladosam** [reply]
@friendly-bits Hi,  yes I'm using it on my VPS server.  And I regulary update it :D It works just fine. My country is not blocked and all others are so thanks for that. 🥇 
I can post info about progress on you're behalf. Just send me text.

**Reply by friendly-bits** [reply]
There was a period where I would release an update every day. Just wanted to implement all planned features and get it ready for OpenWrt. Now both goals have been achieved and geoip-shell has been merged into OpenWrt packages repo. So probably updates will become much more scarce now (unless someone requests additional features or I discover some bugs).

I really appreciate your offer, however please don't feel that you are obliged. That said, if you want to tell the reddit people about the project, I think best if you just describe your own experience with it.
